File: Euclid/Game/Item/ItemManager.cs
using Euclid.Storage.Database.Access;
using Euclid.Storage.Database.Data;
using System.Collections.Generic;
using System.Linq;

namespace Euclid.Game
{
    public class ItemManager : ILoadable
    {
        #region Fields

        public static readonly ItemManager Instance = new ItemManager();

        #endregion

        #region Properties

        public Dictionary<int, ItemDefinition> Definitions { get; set; }

        #endregion

        #region Constructors

        public void Load()
        {
            Definitions = ItemDao.GetDefinitions().Select(x => new ItemDefinition(x)).ToDictionary(x => x.Data.Id, x => x);
        }

        #endregion

        #region Public methods

        /// <summary>
        /// Safe method to try and get definition
        /// </summary>
        public ItemDefinition GetDefinition(int definitionId)
        {
            Definitions.TryGetValue(definitionId, out var definition);
            return definition;
        }

        /// <summary>
        /// Try and resolve an item inside a room or persons inventory
        /// </summary>
        /// <param name="itemId">the item GUID</param>
        /// <returns>the tiem</returns>
        public Item ResolveItem(int itemId, ItemData itemData = null)
        {
            if (itemData == null)
                itemData = ItemDao.GetItem(itemId);

            if (itemData == null)
                return null;

            var room = RoomManager.Instance.GetRoom(itemData.RoomId);

            if (room == null)
            {
                var player = PlayerManager.Instance.GetPlayerById(itemData.RoomId);

                if (player != null)
                {
                    return null;// player.Inventory.GetItem(itemData.Id.ToString());
                }
            }
            else
            {
                if (room.EntityManager.GetEntities<Item>().Count == 0)
                    room.ItemManager.Load();

                return room.ItemManager.GetItem(itemData.Id);
            }


            return null;
        }

        #endregion
    }
}

File: euclid/src/main/java/euclid/game/item/ItemManager.java
package euclid.game.item;

import euclid.game.ILoadable;
import euclid.game.room.Room;
import euclid.game.room.RoomManager;
import euclid.storage.database.access.ItemDao;
import euclid.storage.database.data.ItemData;
import euclid.storage.database.data.ItemDefinitionData;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ItemManager implements ILoadable {

    public static final ItemManager INSTANCE = new ItemManager();

    private Map<Integer, ItemDefinition> definitions;

    private ItemManager() {
    }

    public static ItemManager getInstance() {
        return INSTANCE;
    }

    @Override
    public void load() {
        definitions = new HashMap<>();
        List<ItemDefinitionData> data = ItemDao.getDefinitions();
        for (ItemDefinitionData defData : data) {
            ItemDefinition def = new ItemDefinition(defData);
            definitions.put(defData.getId(), def);
        }
    }

    public Map<Integer, ItemDefinition> getDefinitions() {
        return definitions;
    }

    public ItemDefinition getDefinition(int definitionId) {
        return definitions.get(definitionId);
    }

    public Item resolveItem(int itemId, ItemData itemData) {
        if (itemData == null)
            itemData = ItemDao.getItem(itemId);

        if (itemData == null)
            return null;

        Room room = RoomManager.getInstance().getRoom(itemData.getRoomId());

        if (room == null) {
            return null;
        } else {
            if (room.getEntityManager().getEntities(Item.class).isEmpty())
                room.getItemManager().load();
            return room.getItemManager().getItem(itemData.getId());
        }
    }

    public Item resolveItem(int itemId) {
        return resolveItem(itemId, null);
    }
}
